cdc/capture/http_handler.go

```diff
@@ -0,0 +1,544 @@
+// @Tags changefeed
+// @Accept  json
+// @Produce  json
+// @Param   state  query  string  false  "state"
+// @Success 200 {array} model.ChangefeedCommonInfo
+// @Failure 500 {object} model.HTTPError
+// @Router	/api/v1/changefeeds [get]
+func (h *httpHandler) ListChangefeed(c *gin.Context) {
+	state := c.Query(APIOpVarChangefeedState)
+	statuses, err := h.capture.etcdClient.GetAllChangeFeedStatus(c.Request.Context())
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	changefeedIDs := make(map[string]struct{}, len(statuses))
+	for cid := range statuses {
+		changefeedIDs[cid] = struct{}{}
+	}
+
+	resps := make([]*model.ChangefeedCommonInfo, 0)
+	for changefeedID := range changefeedIDs {
+		cfInfo, err := h.capture.etcdClient.GetChangeFeedInfo(c.Request.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+			return
+		}
+
+		if !httputil.IsFiltered(state, cfInfo.State) {
+			continue
+		}
+
+		cfStatus, _, err := h.capture.etcdClient.GetChangeFeedStatus(c.Request.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+			return
+		}
+
+		resp := &model.ChangefeedCommonInfo{
+			ID: changefeedID,
+		}
+
+		if cfInfo != nil {
+			resp.FeedState = cfInfo.State
+			resp.RunningError = cfInfo.Error
+		}
+
+		if cfStatus != nil {
+			resp.CheckpointTSO = cfStatus.CheckpointTs
+			tm := oracle.GetTimeFromTS(cfStatus.CheckpointTs)
+			resp.CheckpointTime = model.JSONTime(tm)
+		}
+
+		resps = append(resps, resp)
+	}
+	c.IndentedJSON(http.StatusOK, resps)
+}
+
+// GetChangefeed get detail info of a changefeed
+// @Summary Get changefeed
+// @Description get detail information of a changefeed
+// @Tags changefeed
+// @Accept  json
+// @Produce  json
+// @Param   changefeed_id  path  string  true  "changefeed_id"
+// @Success 200 {object} model.ChangefeedDetail
+// @Failure 500,400 {object} model.HTTPError
+// @Router	/api/v1/changefeeds/{changefeed_id} [get]
+func (h *httpHandler) GetChangefeed(c *gin.Context) {
+	changefeedID := c.Param(APIOpVarChangefeedID)
+	if err := model.ValidateChangefeedID(changefeedID); err != nil {
+		c.IndentedJSON(http.StatusBadRequest,
+			model.NewHTTPError(cerror.ErrAPIInvalidParam.GenWithStack("invalid changefeed_id: %s", changefeedID)))
+		return
+	}
+
+	info, err := h.capture.etcdClient.GetChangeFeedInfo(c, changefeedID)
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	status, _, err := h.capture.etcdClient.GetChangeFeedStatus(c, changefeedID)
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	processorInfos, err := h.capture.etcdClient.GetAllTaskStatus(c, changefeedID)
```

Comment:
those info can't be gotten from the memory?

Reply:
yeh